tsdb/mapper.go

```diff
@@ -254,17 +260,43 @@ func (lm *LocalMapper) Open() error {
 	return nil
 }
 
+func (lm *LocalMapper) SetRemote(m Mapper) error {
+	lm.remote = m
+	return nil
+}
+
 func (lm *LocalMapper) NextChunk() (interface{}, error) {
+	// If set, use remote mapper.
+	if lm.remote != nil {
+		b, err := lm.remote.NextChunk()
+		if err != nil {
+			return nil, err
+		} else if b == nil {
+			return nil, nil
+		}
+
+		mo := &MapperOutput{}
+		if err := json.Unmarshal(b.([]byte), mo); err != nil {
+			return nil, err
+		} else if len(mo.Values) == 0 {
+			// Mapper on other node sent 0 values so it's done.
+			return nil, nil
+		}
+		return mo, nil
+	}
+
+	// Remote mapper not set so get values from local shard.
 	if lm.rawMode {
 		return lm.nextChunkRaw()
 	}
+
 	return lm.nextChunkAgg()
 }
 
 // nextChunkRaw returns the next chunk of data. Data comes in the same order as the
 // tags return by TagSets. A chunk never contains data for more than 1 tagset.
 // If there is no more data for any tagset, nil will be returned.
-func (lm *LocalMapper) nextChunkRaw() (*MapperOutput, error) {
+func (lm *LocalMapper) nextChunkRaw() (interface{}, error) {
```

Comment:
Returning `*MapperOutput` instead of `interface{}` here and below in `nextChunkAgg()` caused the loop in `processMapShardRequest` (in cluster/service.go) to run in a tight loop spewing responses because the `if chunk == nil` check was never `true`.  http://play.golang.org/p/aIsHR9UXhf